source/opt/folding_rules.cpp

```diff
@@ -2507,24 +2507,16 @@ FoldingRule RedundantFMix() {
   };
 }
 
-// This rule handles addition of zero for integers.
-FoldingRule RedundantIAdd() {
-  return [](IRContext* context, Instruction* inst,
+FoldingRule RedundantBinaryLRhs0(spv::Op op, size_t arg) {
+  return [op, arg](IRContext* context, Instruction* inst,
             const std::vector<const analysis::Constant*>& constants) {
-    assert(inst->opcode() == spv::Op::OpIAdd &&
-           "Wrong opcode. Should be OpIAdd.");
+    assert(inst->opcode() == op && "Wrong opcode.");
+    assert(constants.size() == 2);
 
-    uint32_t operand = std::numeric_limits<uint32_t>::max();
-    const analysis::Type* operand_type = nullptr;
-    if (constants[0] && constants[0]->IsZero()) {
-      operand = inst->GetSingleWordInOperand(1);
-      operand_type = constants[0]->type();
-    } else if (constants[1] && constants[1]->IsZero()) {
-      operand = inst->GetSingleWordInOperand(0);
-      operand_type = constants[1]->type();
-    }
+    if (constants[arg] && constants[arg]->IsZero()) {
+      auto operand = inst->GetSingleWordInOperand(arg);
```

Comment:
This look like you will always return the operand whose value is 0.
```suggestion
      auto operand = inst->GetSingleWordInOperand(1-arg);
```

Reply:
Oh my god I missed that the original uses the opposite argument in one of these; that was the test failure, yeah.